File: backend/go-app/repository/repository.go
package repository

import "gorm.io/gorm"

type Repository struct {
	db *gorm.DB
}

func (self *Repository) GetUsers() []User {
	var users []User
	self.db.Find(&users)
	return users
}

func (self *Repository) GetTasks() []Task {
	var tasks []Task
	self.db.Model(&Task{}).Preload("User").Where("status != 'deleted'").Find(&tasks)
	return tasks
}

func (self *Repository) CreateTask(task Task) Task {
	self.db.Create(&task)
	return task
}

func(self *Repository) UpdateTask(id int, task Task) Task {
	var res Task
	self.db.First(&res, id).UpdateColumns(task)
	return res
}

func (self *Repository) DeleteTask(id int) Task {
	var task Task
	self.db.First(&task, id).UpdateColumn("status", "deleted")
	return task
}

func GetInstance() (*Repository, error) {
	db, err := GetDb()

	if err != nil {
		return nil, err
	}

	return &Repository{
		db: db,
	}, nil
}

File: backend/elixir_app/lib/elixir_app/repository.ex
defmodule ElixirApp.Repository do
  def get_users() do
    Postgrex.query!(:postgrex, "SELECT * FROM users", [])
    |> ElixirApp.Helper.postgrex_res_to_json()
  end

  def get_tasks() do
    Postgrex.query!(
      :postgrex,
      """
        SELECT "task".*, json_build_object('id', "user"."id", 'image_url' ,"user"."image_url") as "user" FROM "tasks" as "task"
          LEFT JOIN "users" as "user" on "user"."id" = "task"."user_id"
          WHERE "task"."status" != 'deleted'
          GROUP BY "task"."id", "user"."id"
      """,
      []
    )
    |> ElixirApp.Helper.postgrex_res_to_json()
  end
end
